package huginn;

message GPS{
    required double latitude = 1;
    required double longitude = 2;
    required double altitude = 3;
    required double airspeed = 4;
    required double heading = 5;
}

message Accelerometer{
    required double x = 1;
    required double y = 2;
    required double z = 3;
}

message Gyroscope{
    required double roll_rate = 1;
    required double pitch_rate = 2;
    required double yaw_rate = 3;
}

message Thermometer{
    required double temperature = 1;
}

message PressureSensor{
    required double pressure = 1;
}

message PitotTube{
    required double pressure = 1;
}

message Engine{
    required double thrust = 1;
    required double throttle = 2;
} 

message Controls{
    required double aileron = 1;
    required double elevator = 2;
    required double rudder = 3;
    required double throttle = 4;
}

message INS{
    required double roll = 1;
    required double pitch = 2;
    required double latitude = 3;
    required double longitude = 4;
    required double altitude = 5;
    required double airspeed = 6;
    required double heading = 7;
}

message Accelerations{
    required double x = 1;
    required double y = 2;
    required double z = 3;
    required double p_dot = 4;
    required double q_dot = 5;
    required double r_dot = 6;
    required double u_dot = 7;
    required double v_dot = 8;
    required double w_dot = 9;
    required double gravity = 10;
}

message Velocities{
    required double p = 1;
    required double q = 2;
    required double r = 3;
    required double true_airspeed = 4;
    required double climb_rate = 5;
    required double u = 6;
    required double v = 7;
    required double w = 8;
    required double calibrated_airspeed = 9;
    required double equivalent_airspeed = 10;
    required double ground_speed = 11;
}

message Position{
    required double latitude = 1;
    required double longitude = 2;
    required double altitude = 3;
    required double heading = 4;
}

message Orientation{
    required double phi = 1;
    required double theta = 2;
    required double psi = 3;
}

message Atmosphere{
    required double pressure = 1;
    required double sea_level_pressure = 2;
    required double temperature = 3;
    required double sea_level_temperature = 4;
    required double density = 5;
    required double sea_level_density = 6;
}

message Forces{
    required double x_body = 1;
    required double y_body = 2;
    required double z_body = 3;
    required double x_wind = 4;
    required double y_wind = 5;
    required double z_wind = 6;
    required double x_total = 7;
    required double y_total = 8;
    required double z_total = 9;
}

message SimulatorData{
    required double time = 1;
    required GPS gps = 2;
    required Accelerometer accelerometer = 3;
    required Gyroscope gyroscope = 4;
    required Thermometer thermometer = 5;
    required PressureSensor pressure_sensor = 6;
    required PitotTube pitot_tube = 7;
    required Engine engine = 8;
    required Controls controls = 9;
    required INS ins = 10;
    required Accelerations accelerations = 11;
    required Velocities velocities = 12;
    required Position position = 13;
    required Orientation orientation = 14;
    required Atmosphere atmosphere = 15;
    required Forces forces = 16;
}

enum SensorDataRequestType{
    INVALID_REQUEST = 0;
    GPS_REQUEST = 1;
    ACCELEROMETER_REQUEST = 2;
    GYROSCOPE_REQUEST = 3;
    THERMOMETER_REQUEST = 4;
    PRESSURE_SENSOR_REQUEST = 5;
    PITOT_TUBE_REQUEST = 6;
    ENGINE_REQUEST = 7;
    CONTROLS_REQUEST = 8;
    INS_REQUEST = 9;
}

message SensorDataRequest{    
    required SensorDataRequestType type = 1;
}

message SensorDataResponse{
    required SensorDataRequestType type = 1;

    optional GPS gps = 2;
    optional Accelerometer accelerometer = 3;
    optional Gyroscope gyroscope = 4;
    optional Thermometer thermometer = 5;
    optional PressureSensor pressure_sensor = 6;
    optional PitotTube pitot_tube = 7;
    optional Engine engine = 8;
    optional Controls controls = 9;
    optional INS ins = 10;
}